packages/react-utilities/src/compose/types.ts

```diff
@@ -113,6 +113,34 @@ export type ComponentState<Shorthands extends ObjectShorthandPropsRecord> = {
   };
 } & Shorthands;
 
+/**
+ * This is part of a hack to infer the element type from a native element *props* type.
+ * The only place the original element is found in a native props type (at least that's workable
+ * for inference) is in the event handlers, so some of the helper types use this event handler
+ * name to infer the original element type.
+ *
+ * Notes:
+ * - Using an extremely obscure event handler reduces the likelihood that its signature will be
+ *   modified in any component's props.
+ * - Inferring based on a single prop name instead of a larger type like `DOMAttributes<T>` should be
+ *   less expensive for typescript to evaluate and is less likely to result in type expansion in .d.ts.
+ */
+type SomeEventName = 'onLostPointerCaptureCapture';
+
+/**
+ * Return type for `React.forwardRef`, including inference of the proper typing for the ref.
+ */
+export type ForwardRefComponent<Props> = SomeEventName extends keyof Props
```

Comment:
Could we possibly define the type like this? `ForwardRefComponent<Props = {}, RefElement extends Element = Element>`. (this is just an example, I am sure the `RefElement` type would have to be better specified to make React happy)

This would allow us to infer the RefElement type if none was passed in as we are doing currently, but also allow easier type casting for polymorphic components.

Example might look like this:
```tsx
type Props = {
  as?: "input" | "textarea" // very simplified example
  value: string
}

const InputOrTextArea: ForwardRefComponent<Props, HTMLInputElement | HTMLTextAreaElement> = React.forwardRef(...)
```

This can then be better inferred later on by the end user.

Reply:
Maybe `RefElement extends Element = unknown` and then go to full inference if `RefElement extends unknown`? Though I'm still not 100% sure that will fix the type mismatch.